### PR TITLE
refactor(FriendlyDateRanges) Rewrite solution and test cases

### PR DESCRIPTION
**Pull request by S1ngS1ng**
- [x] LGTM
- [ ] LGTM

## Issue link
- #247 
- #173 

## In this PR
- Add corner case description
- Rewrite solution
- Rewrite test cases

## Drawback
- `CurrentYear` is hard coded

### w/ Current Year

| Description | Test input | Output |
| --- | --- | --- |
| Same month | `['2017-01-02', '2017-01-05']` | `['January 2nd', '5th']` |
| Different month | `['2017-02-01', '2017-03-03']` | `['February 1st', 'March 3rd']` |
| End with current year, `yearDiff < 1` | `['2016-05-11', '2017-04-04']` | `['May 11th, 2016', 'April 4th']` |
| Start with current year, `yearDiff < 1` | `['2017-07-12', '2018-06-13']` | `['July 12th', 'June 13th']` |

### Year diff
| Description | Test input | Output |
| --- | --- | --- |
| `yearDiff > 1` w/o current year | `['2003-08-15', '2009-09-21']` | `['August 15th, 2003', 'September 21st, 2009']` |
| `yearDiff === 1` && `daysDiff < 365` | `['2010-10-23', '2011-10-22']` | `['October 23rd, 2010', 'October 22nd']` |
| `yearDiff === 1` && `daysDiff === 365` | `['2008-10-31', '2009-10-31']` | `['October 31st, 2008', 'October 31st, 2009']` |
| `yearDiff === 1` && `daysDiff > 365` | `['2004-11-17', '2005-12-25']` | `['November 17th, 2004', 'December 25th, 2005']` |

### Other
| Description | Test input | Output |
| --- | --- | --- |
| Identical dates | `['2001-12-20', '2001-12-20']` | `['December 20th, 2001']` |
| Corner case (Invalid range) | `['2002-12-20', '2001-12-20']` | `undefined` |

### Also included
- ALL months string
- Ordinal numbers test

| Input | Output |
| --- | --- |
| `1, 21, 31` | ends with `st` |
| `2, 22` | ends with `nd` |
| `3, 23` | ends with `rd` |
| `11, 12, 13` | ends with `th` |
| All other | ends with `th` |

### CODE REVIEW

File: seed/challenges/01-front-end-development-certification/advanced-bonfires.json [hunk 1]
```diff
@@ -236,28 +236,32 @@
...
...      date: ordinalDay(date.getDate()),\n            year: date.getFullYear()\n        }\n    }\n    function getResult(firstDate, secondDate) {\n        var parsedFirst = parseDateString(firstDate);\n        var parsedSecond = parseDateString(secondDate);\n        var firstObject = transformDate(firstDate);\n        var secondObject = transformDate(secondDate);\n        if (firstObject.year === currentYear) {\n            firstObject.year = false;\n        }\n        if (getDiff('year', parsedFirst, parsedSecond) === 0) {\n            secondObject.year = false;\n            if (getDiff('month', parsedFirst, parsedSecond) === 0) {\n                secondObject.month = false;\n            }\n            if (getDiff('date', parsedFirst, parsedSecond) === 0) {\n                secondObject.date = false;\n            }\n        } else if (getDiff('year', parsedFirst, parsedSecond) === 1) {\n            if (getDiff('unixTime', parsedFirst, parsedSecond) / (1000 * 3600 * 24) < 365) {\n                secondObject.year = false;\n            }\n        }\n        return [].concat.apply(composeOutput(firstObject), composeOutput(secondObject))\n    }\n    return getResult(arr[0], arr[1])\n}"
```
Comment (S1ngS1ng) [inline seed/challenges/01-front-end-development-certification/advanced-bonfires.json:252]: ~~Possible optimization: `var dateArr = dateStr.split('-'); return new Date(dateArr[0], dateArr[1] - 1, dateArr[2])` => `return new Date(dateStr)`~~

@wudifeixue 
image
Interesting 😂

**Comment by S1ngS1ng** [comment]
@wudifeixue So, either use `getUTCDate` to make it time-zone independent or `split` and initialize date with numbers.
image

So, I'm not gonna make the change mentioned above.

File: seed/challenges/01-front-end-development-certification/advanced-bonfires.json (hunk 1 above)
Comment (wudifeixue) [reply seed/challenges/01-front-end-development-certification/advanced-bonfires.json:252]: Hmm, interesting indeed. I saw it during work but I guess this is part of won't fix?

Comment (S1ngS1ng) [reply seed/challenges/01-front-end-development-certification/advanced-bonfires.json:252]: @wudifeixue It is actually not "won't fix". All the code here still works 😃

Comment (wudifeixue) [inline seed/challenges/01-front-end-development-certification/advanced-bonfires.json:252]: 唉呀妈呀，这段读起来真的是太痛苦了..

Comment (S1ngS1ng) [reply seed/challenges/01-front-end-development-certification/advanced-bonfires.json:252]: @wudifeixue There you go 😄 
```JS
function makeFriendlyDates(arr) {
    if (getDiff('unixTime', parseDateString(arr[0]), parseDateString(arr[1])) < 0) {
        return;
    }
    var current = new Date();
    var currentYear = current.getFullYear();
    var monthArr = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
    function ordinalDay(day){
        if (day !== 11 && day.toString().slice(-1) === '1') {
            return day + 'st';
        } else if (day !== 12 && day.toString().slice(-1) === '2') {
            return day + 'nd';
        } else if (day !== 13 && day.toString().slice(-1) === '3') {
            return day + 'rd';
        }
        return day + 'th';
    }
    function getDiff(attr, firstDate, secondDate) {
        var attrMap = {
            year: 'getFullYear',
            month: 'getMonth',
            date: 'getDate',
            unixTime: 'getTime'
        }
        return secondDate[attrMap[attr]]() - firstDate[attrMap[attr]]();
    }
    function parseDateString(dateStr) {
        var dateArr = dateStr.split('-');
        return new Date(dateArr[0], dateArr[1] - 1, dateArr[2]);
    }
    function composeOutput(dateObj) {
        var output = '';
        if (dateObj.month) {
            output += dateObj.month;
        }
        if (dateObj.date) {
            output += (output ? ' ' : '') + dateObj.date;
        }
        if (dateObj.year) {
            output += ', ' + dateObj.year;
        }
        return output ? [output] : undefined;
    }
    function transformDate(dateString) {
        var date = parseDateString(dateString);
        return {
            month: monthArr[date.getMonth()],
            date: ordinalDay(date.getDate()),
            year: date.getFullYear()
        }
    }
    function getResult(firstDate, secondDate) {
        var parsedFirst = parseDateString(firstDate);
        var parsedSecond = parseDateString(secondDate);
        var firstObject = transformDate(firstDate);
        var secondObject = transformDate(secondDate);
        if (firstObject.year === currentYear) {
            firstObject.year = false;
        }
        if (getDiff('year', parsedFirst, parsedSecond) === 0) {
            secondObject.year = false;
            if (getDiff('month', parsedFirst, parsedSecond) === 0) {
                secondObject.month = false;
            }
            if (getDiff('date', parsedFirst, parsedSecond) === 0) {
                secondObject.date = false;
            }
        } else if (getDiff('year', parsedFirst, parsedSecond) === 1) {
            if (getDiff('unixTime', parsedFirst, parsedSecond) / (1000 * 3600 * 24) < 365) {
                secondObject.year = false;
            }
        }
        return [].concat.apply(composeOutput(firstObject), composeOutput(secondObject))
    }
    return getResult(arr[0], arr[1])
}
```

Comment (wudifeixue) [reply seed/challenges/01-front-end-development-certification/advanced-bonfires.json:252]: aha!

File: seed/challenges/01-front-end-development-certification/advanced-bonfires.json [hunk 2]
```diff
@@ -236,28 +236,32 @@
...
+        "assert.deepEqual(makeFriendlyDates(['2010-10-23', '2011-10-22']), ['October 23rd, 2010', 'October 22nd'], 'message: <code>makeFriendlyDates([\"2010-10-23\", \"2011-10-22\"])</code> 应该返回 <code>[\"October 23rd, 2010\",\"October 22nd\"]</code>.');",
+        "assert.deepEqual(makeFriendlyDates(['2008-10-31', '2009-10-31']), ['October 31st, 2008', 'October 31st, 2009'], 'message: <code>makeFriendlyDates([\"2008-10-31\", \"2009-10-31\"])</code> 应该返回 <code>[\"October 31st, 2008\",\"October 31st, 2009\"]</code>.');",
+        "assert.deepEqual(makeFriendlyDates(['2004-11-17', '2005-12-25']), ['November 17th, 2004', 'December 25th, 2005'], 'message: <code>makeFriendlyDates([\"2004-11-17\", \"2005-12-25\"])</code> 应该返回 <code>[\"November 17th, 2004\",\"December 25th, 2005\"]</code>.');",
+        "assert.deepEqual(makeFriendlyDates(['2001-12-20', '2001-12-20']), ['December 20th, 2001'], 'message: <code>makeFriendlyDates([\"2001-12-20\", \"2001-12-20\"])</code> 应该返回 <code>[\"December 20th, 2001\"]</code>.');",
+        "assert.isNotArray(makeFriendlyDates(['2002-12-20', '2001-12-20']), 'message: <code>makeFriendlyDates([\"2001-12-20\", \"2001-12-20\"])</code> 不应该返回数组');"
```
Comment (wudifeixue) [inline seed/challenges/01-front-end-development-certification/advanced-bonfires.json:264]: What is this assert.isNotArray? It seems to be the root cause of npm test not passing.

**Review by wudifeixue — COMMENTED** [review]
npm test would not pass
